sdks/java/io/exec/src/main/java/org/apache/beam/sdk/io/exec/ExecIO.java

```diff
@@ -0,0 +1,325 @@
+      abstract Write build();
+    }
+
+    public Write withWorkingDir(String workingDir) {
+      checkArgument(workingDir != null,
+          "ExecIO.write().withWorkingDir(workingDir) called with null workingDir");
+      return builder()
+          .setExecOptions(ExecOptions.create(workingDir, getExecOptions().getTimeout()))
+          .build();
+    }
+
+    public Write withTimeout(long timeout) {
+      checkArgument(timeout >= 0,
+          "ExecIO.write().withTimeout(timeout) called with invalid timeout");
+      return builder()
+          .setExecOptions(ExecOptions.create(getExecOptions().getWorkingDir(), timeout))
+          .build();
+    }
+
+    @Override
+    public PDone apply(PCollection<String> input) {
+      input.apply(ParDo.of(new ExecFn(getExecOptions())));
+      return PDone.in(input.getPipeline());
+    }
+
+    @Override
+    public void validate(PCollection<String> input) {
+      // nothing to do
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      builder.addIfNotNull(DisplayData.item("workingDir", getExecOptions().getWorkingDir()));
+      builder.add(DisplayData.item("timeout", getExecOptions().getTimeout()));
+    }
+
+  }
+
+  /**
+   * A POJO describing system command execution options.
+   */
+  @AutoValue
+  public abstract static class ExecOptions implements Serializable {
+
+    @Nullable abstract String getWorkingDir();
+    abstract long getTimeout();
+
+    public static ExecOptions create() {
+      return new AutoValue_ExecIO_ExecOptions(null, Long.MAX_VALUE);
+    }
+
+    public static ExecOptions create(String workingDir) {
+      return new AutoValue_ExecIO_ExecOptions(workingDir, Long.MAX_VALUE);
+    }
+
+    public static ExecOptions create(long timeout) {
+      return new AutoValue_ExecIO_ExecOptions(null, timeout);
+    }
+
+    public static ExecOptions create(String workingDir, long timeout) {
+      return new AutoValue_ExecIO_ExecOptions(workingDir, timeout);
+    }
+
+  }
+
+  /**
+   * Main {@link DoFn} which actually executes system commands contained in the input
+   * {@link PCollection}.
+   * The output {@link PCollection} contains the commands execution output.
+   * This function optionally accepts {@link ExecOptions} as constructor argument.
+   */
+  public static class ExecFn extends DoFn<String, String> {
+
+
+    private ExecOptions options;
+
+    public ExecFn() {
+      this.options = ExecOptions.create();
+    }
+
+    public ExecFn(ExecOptions options) {
+      this.options = options;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      String command = context.element();
+
+      checkState(command != null, "ExecIO executed with null command");
+
+      ByteArrayOutputStream out = new ByteArrayOutputStream();
+
+      DefaultExecutor executor = new DefaultExecutor();
+      PumpStreamHandler handler = new PumpStreamHandler(out, out);
```

Comment:
One should be able to control whether one gets the stderr or stdout, or possibly get a tuple rather than an interleaved string.